### PR TITLE
bitrot: Add bitrot plugin to glusterd2

### PR DESCRIPTION
**Pull request by kotreshhr**
This patch adds following support
1. Enable bitrot : Starts bitd and scrubber
2. Disable bitrot: Stops bitd and scrubber

Signed-off-by: Kotresh HR <khiremat@redhat.com>

### CODE REVIEW

**Comment by prashanthpai** [comment]
add to whitelist

File: plugins/bitrot/transactions.go [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+		return err
+	}
+
+	// Volume is started
+	if volinfo.State != volume.VolStarted {
+		return errors.New("Volume is not started")
+	}
+
+	// bitrot already enabled
+	if volinfo.BitrotEnabled {
```
Comment (aravindavk) [inline plugins/bitrot/transactions.go]: Move all these steps to handler function itself. This Transaction step is executed in all the nodes.

File: plugins/bitrot/transactions.go [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+		return errors.New("Bitrot already enabled on the volume")
+	}
+
+	// Set bitd status in volinfo
+	volinfo.BitrotEnabled = true
```
Comment (aravindavk) [inline plugins/bitrot/transactions.go]: Same as above. Change Volinfo only in handler before calling `Ctx.Set`

File: plugins/bitrot/transactions.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+	// Set scrub status in volinfo
+	volinfo.BitrotScrubEnabled = true
+
+	err := volume.AddOrUpdateVolumeFunc(&volinfo)
```
Comment (aravindavk) [inline plugins/bitrot/transactions.go]: Same  as above.

File: plugins/bitrot/transactions.go (hunk 1 above)
Comment (kotreshhr) [reply plugins/bitrot/transactions.go]: Done

File: plugins/bitrot/transactions.go (hunk 2 above)
Comment (kotreshhr) [reply plugins/bitrot/transactions.go]: Done

**Comment by kotreshhr** [comment]
Addressed all the issues. Now it the bitrot enable/disable works fine.

File: plugins/bitrot/init.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+// RestRoutes returns list of REST API routes to register with Glusterd
+func (p *Plugin) RestRoutes() route.Routes {
+	return route.Routes{
+		route.Route{
+			Name:        "BitrotHelp",
+			Method:      "GET",
+			Pattern:     "/bitrot/help",
+			Version:     1,
+			HandlerFunc: bitrotHelpHandler},
```
Comment (aravindavk) [inline plugins/bitrot/init.go]: This route can be removed, added just for example route

File: plugins/bitrot/rest.go [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+	log "github.com/sirupsen/logrus"
+
+	"github.com/pborman/uuid"
+)
+
+func bitrotHelpHandler(w http.ResponseWriter, r *http.Request) {
+	// Implement the help logic and send response back as below
+	ctx := r.Context()
+	restutils.SendHTTPResponse(ctx, w, http.StatusOK, "Bitrot Help")
+}
```
Comment (aravindavk) [inline plugins/bitrot/rest.go]: Same as above, this can be removed

File: plugins/bitrot/rest.go [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+	}
+
+	// Check if volume is started
+	if volinfo.State != volume.VolStarted {
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, errors.ErrVolNotStarted.Error(), api.ErrCodeDefault)
+		return
+	}
+
+	// Check if bitrot is already enabled
+	if volinfo.BitrotEnabled {
```
Comment (aravindavk) [inline plugins/bitrot/rest.go]: Can be changed to check volinfo.Options[VkeyFeaturesBitrot]

    val, ok := volinfo.Options[VkeyFeaturesBitrot]
    if ok && val == "on" {

File: plugins/bitrot/rest.go [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+
+	// Validate volume existence
+	volinfo, err := volume.GetVolume(volName)
+	if err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusNotFound, errors.ErrVolNotFound.Error(), api.ErrCodeDefault)
+		return
+	}
+
+	// Check if bitrot is already disabled
+	if !volinfo.BitrotEnabled {
```
Comment (aravindavk) [inline plugins/bitrot/rest.go]: Same as above

File: plugins/bitrot/transactions.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+// IsBitrotAffectedNode returns true if there are local bricks of volume on which bitrot is enabled
+func IsBitrotAffectedNode() bool {
+	volumes, e := volume.GetVolumes()
+	if e != nil {
+		log.WithError(e).Error("Failed to get volumes")
+		return true
+	}
+	for _, v := range volumes {
+		if !v.BitrotEnabled {
```
Comment (aravindavk) [inline plugins/bitrot/transactions.go]: Change required here to check from Options

File: glusterd2/volume/struct.go [hunk 9]
```diff
@@ -85,18 +85,19 @@ type Subvol struct {
...
+	Transport     string
+	DistCount     int
+	Options       map[string]string
+	State         VolState
+	Checksum      uint64
+	Version       uint64
+	Subvols       []Subvol
+	Auth          VolAuth // TODO: should not be returned to client
+	GraphMap      map[string]string
+	BitrotEnabled bool
```
Comment (aravindavk) [inline glusterd2/volume/struct.go]: This can be removed.

Comment (kotreshhr) [reply glusterd2/volume/struct.go]: Done

File: plugins/bitrot/init.go (hunk 4 above)
Comment (kotreshhr) [reply plugins/bitrot/init.go]: Done

**Comment by kotreshhr** [comment]
All the review comments are addressed

**Pull request merged**